File: numberOf1InBinary.js
//法一
function NumberOf1_1(n)
{
    // n每次和1做与运算，1每次左移。一共左移次数为n的位数
    var count=0;
    var flag=1;
    while(flag){
        if(n&flag){
            count+=1;
        }
        flag=flag<<1;
    }
    return count;
}

//法二
function NumberOf1_2(n)
{
    //n减去1，再和原整数做与运算，把该整数最右边的1变成0.二进制中有几个1进行几次这样的运算。循环的次数为n中1的的个数。
    var count=0;
    while(n){
        count+=1;
        n=(n-1)&n;
    }
    return count;
}
console.log(NumberOf1_2(9));

File: numberOf1InBinary.py
# -*- coding:utf-8 -*-
"""
思路： 
暴力解决：分为两种情况：一种整数n>=0,直接化为二进制.一种n<0，求其补码
其中求补码的方法为将其用原码表示，然后从右边第一个为‘1’的数起往左依次取反（取反操作即与1异或，0^1=1,1^1=0）
"""
class Solution:
    def NumberOf1(self, n):
        # write code here
        if n >= 0:
            num =  ('{:032b}'.format(n))
        else:
            res =  ('{:032b}'.format(-n))[1:]
            num =  '1'
            flag = 0
            for i in range(len(res)-1,-1,-1):
                if res[i] == '1':
                    index = i
                    flag = 1
                    break
            if flag:
                for i in res[:index]:
                    num += str(int(i) ^ 1)
                num += res[index:]
            else:
                num += res
            print(num,len(num))
        return num.count('1')

if __name__ == "__main__":
    print(Solution().NumberOf1(9))
    
